src/v/coproc/script_context.cc

```diff
@@ -0,0 +1,335 @@
+              return process_one_reply(std::move(e));
+          });
+    });
+}
+
+ss::future<> script_context::process_one_reply(process_batch_reply::data e) {
+    /// Ensure this 'script_context' instance is handling the correct reply
+    if (e.id != _id) {
+        return ss::make_exception_future<>(std::logic_error(fmt::format(
+          "id: {} recieved response from another script_id: {}", _id, e.id)));
+    }
+    /// Use the source topic portion of the materialized topic to perform a
+    /// lookup for the relevent 'ntp_context'
+    auto materialized_ntp = model::materialized_ntp(e.ntp);
+    auto found = _ntp_ctxs.find(materialized_ntp.source_ntp());
+    if (found == _ntp_ctxs.end()) {
+        vlog(
+          coproclog.warn,
+          "script {} unknown source ntp: {}",
+          _id,
+          materialized_ntp.source_ntp());
+        return ss::now();
+    }
+    auto ntp_ctx = found->second;
+    return write_materialized(materialized_ntp, std::move(e.reader))
+      .then([this, ntp_ctx](bool crc_parse_failure) {
+          if (crc_parse_failure) {
+              vlog(coproclog.warn, "record_batch failed to pass crc checks");
+              return;
+          }
+          auto ofound = ntp_ctx->offsets.find(_id);
+          vassert(
+            ofound != ntp_ctx->offsets.end(),
+            "Offset not found for script id {} for ntp owning context: {}",
+            _id,
+            ntp_ctx->ntp());
+          /// Reset the acked offset so that progress can be made
+          ofound->second.last_acked = ofound->second.last_read;
+      });
+}
+
+ss::future<storage::log> get_log(storage::api& api, const model::ntp& ntp) {
+    auto found = api.log_mgr().get(ntp);
+    if (found) {
+        return ss::make_ready_future<storage::log>(*found);
+    }
+    vlog(coproclog.info, "Making new log: {}", ntp);
+    return api.log_mgr().manage(
+      storage::ntp_config(ntp, api.log_mgr().config().base_dir));
+}
+
+ss::future<bool>
+write_checked(storage::log log, model::record_batch_reader reader) {
+    const storage::log_append_config write_cfg{
+      .should_fsync = storage::log_append_config::fsync::no,
+      .io_priority = ss::default_priority_class(),
+      .timeout = model::no_timeout};
+    return std::move(reader)
+      .for_each_ref(
+        coproc::reference_window_consumer(
+          model::record_batch_crc_checker(), log.make_appender(write_cfg)),
+        model::no_timeout)
+      .then([](std::tuple<bool, ss::future<storage::append_result>> t) {
+          const auto& [crc_parse_success, _] = t;
+          return !crc_parse_success;
+      });
+}
+
+ss::future<bool> script_context::write_materialized(
+  const model::materialized_ntp& m_ntp, model::record_batch_reader reader) {
+    auto found = _resources.log_mtx.find(m_ntp.input_ntp());
+    if (found == _resources.log_mtx.end()) {
+        found = _resources.log_mtx.emplace(m_ntp.input_ntp(), mutex()).first;
+    }
+    return found->second.with(
```

Comment:
Hi reviewers, here is the modification to solve #380. Not sure if this is too heavy of a solution, but it involves a map of destination ntps to mutexs. Without this, there are races within async work provided to the underlying log and the effect is different assertions thrown in log.h or segment.h.

Reply:
`are races within async work provided to the underlying log` what does this mean?